mempool/clist_mempool.go

```diff
@@ -0,0 +1,780 @@
+	mem.updateMtx.RLock()
+	defer mem.updateMtx.RUnlock()
+
+	if max < 0 {
+		max = mem.txs.Len()
+	}
+
+	txs := make([]types.Tx, 0, tmmath.MinInt(mem.txs.Len(), max))
+	for e := mem.txs.Front(); e != nil && len(txs) <= max; e = e.Next() {
+		memTx := e.Value.(*MempoolTx)
+		txs = append(txs, memTx.Tx)
+	}
+	return txs
+}
+
+// Lock() must be help by the caller during execution.
+func (mem *CListMempool) Update(
+	height int64,
+	txs types.Txs,
+	deliverTxResponses []*abci.ResponseDeliverTx,
+	preCheck PreCheckFunc,
+	postCheck PostCheckFunc,
+) error {
+	// Set height
+	mem.height = height
+	mem.notifiedTxsAvailable = false
+
+	if preCheck != nil {
+		mem.preCheck = preCheck
+	}
+	if postCheck != nil {
+		mem.postCheck = postCheck
+	}
+
+	for i, tx := range txs {
+		if deliverTxResponses[i].Code == abci.CodeTypeOK {
+			// Add valid committed tx to the cache (if missing).
+			_ = mem.cache.Push(tx)
+		} else {
+			// Allow invalid transactions to be resubmitted.
+			mem.cache.Remove(tx)
+		}
+
+		// Remove committed tx from the mempool.
+		//
+		// Note an evil proposer can drop valid txs!
+		// Mempool before:
+		//   100 -> 101 -> 102
+		// Block, proposed by an evil proposer:
+		//   101 -> 102
+		// Mempool after:
+		//   100
+		// https://github.com/tendermint/tendermint/issues/3322.
+		if e, ok := mem.txsMap.Load(TxKey(tx)); ok {
+			mem.removeTx(tx, e.(*clist.CElement), false)
+		}
+	}
+
+	// Either recheck non-committed txs to see if they became invalid
+	// or just notify there're some txs left.
+	if mem.Size() > 0 {
+		if mem.config.Recheck {
+			mem.logger.Info("Recheck txs", "numtxs", mem.Size(), "height", height)
+			mem.recheckTxs()
+			// At this point, mem.txs are being rechecked.
+			// mem.recheckCursor re-scans mem.txs and possibly removes some txs.
+			// Before mem.Reap(), we should wait for mem.recheckCursor to be nil.
+		} else {
+			mem.notifyTxsAvailable()
+		}
+	}
+
+	// Update metrics
+	mem.metrics.Size.Set(float64(mem.Size()))
+
+	return nil
+}
+
+func (mem *CListMempool) recheckTxs() {
+	if mem.Size() == 0 {
+		panic("recheckTxs is called, but the mempool is empty")
+	}
+
+	mem.recheckCursor = mem.txs.Front()
+	mem.recheckEnd = mem.txs.Back()
+
+	ctx := context.Background()
+
+	// Push txs to proxyAppConn
+	// NOTE: globalCb may be called concurrently.
+	for e := mem.txs.Front(); e != nil; e = e.Next() {
+		memTx := e.Value.(*MempoolTx)
+		_, err := mem.proxyAppConn.CheckTxAsync(ctx, abci.RequestCheckTx{
+			Tx:   memTx.Tx,
+			Type: abci.CheckTxType_Recheck,
+		})
+		if err != nil {
+			// No need in retrying since memTx will be rechecked after next block.
+			mem.logger.Error("Can't check tx", "err", err)
+		}
+	}
+
+	_, err := mem.proxyAppConn.FlushAsync(ctx)
+	if err != nil {
+		mem.logger.Error("Can't flush txs", "err", err)
+	}
+}
+
+//--------------------------------------------------------------------------------
+
+// MempoolTx is a transaction that successfully ran
+type MempoolTx struct {
```

Comment:
That is now exported should be the only change from the tendermint mempool, right?

Reply:
Yes, this is the only change. I won't squash during a merge this time, to make this change more visible.

Reply:
Makes sense! Thanks.